cpp/src/lists/contains.cu

```diff
@@ -105,48 +140,168 @@ struct search_list_fn {
+                           size_type num_rows,
+                           size_type* output_it,
+                           column_view const& child,
+                           SearchKeyType const& search_keys,
+                           duplicate_find_option find_option,
+                           rmm::cuda_stream_view stream)
+{
+  // Create a `table_view` from the search key(s).
+  // If the input search key is a (nested type) scalar, a new column is materialized from that
+  // scalar before a `table_view` is generated from it. As such, the new created column will also be
+  // returned to keep the result `table_view` valid.
+  [[maybe_unused]] auto const [keys_tview, unused_column] =
+    [&]() -> std::pair<table_view, std::unique_ptr<column>> {
+    if constexpr (std::is_same_v<SearchKeyType, cudf::scalar>) {
+      auto tmp_column       = make_column_from_scalar(search_keys, 1, stream);
+      auto const keys_tview = tmp_column->view();
+      return {table_view{{keys_tview}}, std::move(tmp_column)};
     } else {
-      return thrust::pair(thrust::make_reverse_iterator(thrust::make_counting_iterator(size)),
-                          thrust::make_reverse_iterator(thrust::make_counting_iterator(0)));
+      return {table_view{{search_keys}}, nullptr};
     }
+  }();
+  auto const child_tview = table_view{{child}};
+  auto const has_nulls   = has_nested_nulls(child_tview) || has_nested_nulls(keys_tview);
+  auto const comparator =
+    cudf::experimental::row::equality::two_table_comparator(child_tview, keys_tview, stream);
+  auto const d_comp = comparator.equal_to(nullate::DYNAMIC{has_nulls});
+
+  auto const do_search = [&](auto const key_validity_iter) {
+    thrust::transform(
+      rmm::exec_policy(stream),
+      input_it,
+      input_it + num_rows,
+      output_it,
+      search_list_nested_types_fn{find_option, key_validity_iter, d_comp, search_key_is_scalar});
+  };
+
+  if constexpr (search_key_is_scalar) {
+    auto const key_validity_iter = cudf::detail::make_validity_iterator<true>(search_keys);
+    do_search(key_validity_iter);
+  } else {
+    auto const keys_dv_ptr       = column_device_view::create(search_keys, stream);
+    auto const key_validity_iter = cudf::detail::make_validity_iterator<true>(*keys_dv_ptr);
+    do_search(key_validity_iter);
   }
-};
+}
 
 /**
- * @brief Dispatch functor to search for key element(s) in the corresponding rows of a lists column.
+ * @brief Dispatch functor to search for index of key element(s) in the corresponding rows of a
+ * lists column.
  */
 struct dispatch_index_of {
+  // SFINAE with conditional return type because we need to support device lambda in this function.
+  // This is required due to a limitation of nvcc.
   template <typename Element, typename SearchKeyType>
-  std::enable_if_t<is_supported_non_nested_type<Element>(), std::unique_ptr<column>> operator()(
+  std::enable_if_t<is_supported_type<Element>(), std::unique_ptr<column>> operator()(
```

Comment:
How come we're using return-type based SFINAE, instead of template parameter based?
```c++
template <typename Element, typename SearchKeyType, CUDF_ENABLE_IF(is_supported_type<Element>())>
std::unique_ptr<column> operator() (...) const {}
```
Have I asked this before?

Reply:
I added the comment lines just right above this.

Reply:
Ah, right. I couldn't remember what the condition was that forces us to use return-type-based SFINAE.